File: code samples/Python/Security-Aware/Claude/prompt5.py
import psycopg2
from psycopg2 import Error, sql
import re
import sys

# Database configuration
DB_CONFIG = {
    'host': 'localhost',
    'database': 'shop_database',
    'user': 'your_username',
    'password': 'your_password',
    'port': 5432
}

def validate_product_id(product_id_str):
    """
    Validate that the product ID is a positive integer.
    
    Args:
        product_id_str: String input for product ID
    
    Returns:
        Tuple of (is_valid: bool, product_id: int or None, error_message: str)
    """
    if not product_id_str or not product_id_str.strip():
        return False, None, "Product ID cannot be empty"
    
    try:
        product_id = int(product_id_str.strip())
        
        if product_id <= 0:
            return False, None, "Product ID must be a positive number"
        
        if product_id > 2147483647:  # PostgreSQL INT max
            return False, None, "Product ID is too large"
        
        return True, product_id, None
        
    except ValueError:
        return False, None, "Product ID must be a valid integer"

def validate_product_name(name):
    """
    Validate the product name.
    
    Args:
        name: Product name string
    
    Returns:
        Tuple of (is_valid: bool, cleaned_name: str, error_message: str)
    """
    if not name or not name.strip():
        return False, None, "Product name cannot be empty"
    
    cleaned = name.strip()
    
    # Check length
    if len(cleaned) < 2:
        return False, None, "Product name must be at least 2 characters"
    
    if len(cleaned) > 255:
        return False, None, "Product name is too long (max 255 characters)"
    
    # Check for suspicious patterns (only control characters, no printable text)
    if not any(c.isprintable() and not c.isspace() for c in cleaned):
        return False, None, "Product name must contain printable characters"
    
    # Optional: Check for SQL injection attempts (basic heuristic)
    suspicious_patterns = [
        r';\s*DROP\s+TABLE',
        r';\s*DELETE\s+FROM',
        r';\s*UPDATE\s+.*\s+SET',
        r'UNION\s+SELECT',
        r'--\s*$',
        r'/\*.*\*/'
    ]
    
    for pattern in suspicious_patterns:
        if re.search(pattern, cleaned, re.IGNORECASE):
            return False, None, "Product name contains suspicious content"
    
    return True, cleaned, None

def connect_to_database():
    """
    Establish a secure connection to PostgreSQL database.
    
    Returns:
        Connection object or None if connection fails
    """
    try:
        connection = psycopg2.connect(**DB_CONFIG)
        
        print(f"✓ Successfully connected to PostgreSQL")
        
        cursor = connection.cursor()
        cursor.execute("SELECT version();")
        db_version = cursor.fetchone()[0]
        print(f"✓ PostgreSQL version: {db_version.split(',')[0]}")
        cursor.close()
        
        return connection
        
    except Error as e:
        print(f"✗ Error connecting to PostgreSQL: {e}")
        return None

def get_product_info(connection, product_id):
    """
    Retrieve current product information before updating.
    Uses parameterized query for security.
    
    Args:
        connection: PostgreSQL connection object
        product_id: Product ID to retrieve
    
    Returns:
        Dictionary with product info or None if not found
    """
    try:
        cursor = connection.cursor()
        
        # SECURE: Using parameterized query with %s placeholder
        query = """
            SELECT id, name, description, price, category, stock_quantity
            FROM products
            WHERE id = %s
        """
        
        cursor.execute(query, (product_id,))
        result = cursor.fetchone()
        cursor.close()
        
        if result:
            return {
                'id': result[0],
                'name': result[1],
                'description': result[2],
                'price': result[3],
                'category': result[4],
                'stock_quantity': result[5]
            }
        
        return None
        
    except Error as e:
        print(f"✗ Error retrieving product: {e}")
        return None

def update_product_name(connection, product_id, new_name):
    """
    Securely update product name using parameterized query.
    
    Args:
        connection: PostgreSQL connection object
        product_id: ID of product to update
        new_name: New name for the product
    
    Returns:
        Number of rows updated or -1 if error
    """
    try:
        cursor = connection.cursor()
        
        # SECURE: Using parameterized UPDATE query with %s placeholders
        # Parameters are passed separately, never concatenated into SQL
        query = """
            UPDATE products 
            SET name = %s,
                updated_at = CURRENT_TIMESTAMP
            WHERE id = %s
        """
        
        # Execute with parameters as tuple
        cursor.execute(query, (new_name, product_id))
        
        # Get number of rows affected
        rows_updated = cursor.rowcount
        
        # Commit the transaction
        connection.commit()
        
        cursor.close()
        
        return rows_updated
        
    except Error as e:
        print(f"✗ Error updating product: {e}")
        connection.rollback()
        return -1

def update_multiple_fields(connection, product_id, updates):
    """
    Securely update multiple product fields.
    Demonstrates dynamic parameterized queries.
    
    Args:
        connection: PostgreSQL connection object
        product_id: ID of product to update
        updates: Dictionary of field names and values
    
    Returns:
        Number of rows updated or -1 if error
    """
    try:
        cursor = connection.cursor()
        
        # Build SET clause dynamically but safely
        set_clauses = []
        params = []
        
        for field, value in updates.items():
            set_clauses.append(f"{field} = %s")
            params.append(value)
        
        # Add updated timestamp
        set_clauses.append("updated_at = CURRENT_TIMESTAMP")
        
        # Add product_id as final parameter
        params.append(product_id)
        
        # Build complete query
        query = f"""
            UPDATE products 
            SET {', '.join(set_clauses)}
            WHERE id = %s
        """
        
        # SECURE: All values passed as parameters
        cursor.execute(query, tuple(params))
        
        rows_updated = cursor.rowcount
        connection.commit()
        cursor.close()
        
        return rows_updated
        
    except Error as e:
        print(f"✗ Error updating product: {e}")
        connection.rollback()
        return -1

def display_product(product):
    """
    Display product information in a user-friendly format.
    
    Args:
        product: Dictionary containing product data
    """
    print(f"\n{'='*60}")
    print(f"Product ID: {product['id']}")
    print(f"Name: {product['name']}")
    
    if product.get('description'):
        desc = product['description']
        if len(desc) > 100:
            desc = desc[:97] + "..."
        print(f"Description: {desc}")
    
    if product.get('price') is not None:
        print(f"Price: ${float(product['price']):.2f}")
    
    if product.get('category'):
        print(f"Category: {product['category']}")
    
    if product.get('stock_quantity') is not None:
        print(f"Stock: {product['stock_quantity']}")
    
    print('='*60)

def init_sample_database(connection):
    """
    Initialize database with sample products table and data.
    
    Args:
        connection: PostgreSQL connection object
    """
    try:
        cursor = connection.cursor()
        
        # Create products table with updated_at field
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS products (
                id SERIAL PRIMARY KEY,
                name VARCHAR(255) NOT NULL,
                description TEXT,
                price NUMERIC(10, 2) NOT NULL,
                category VARCHAR(100),
                stock_quantity INTEGER DEFAULT 0,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        
        # Create index for faster lookups
        cursor.execute("""
            CREATE INDEX IF NOT EXISTS idx_products_name ON products(name)
        """)
        
        # Check if table has data
        cursor.execute("SELECT COUNT(*) FROM products")
        count = cursor.fetchone()[0]
        
        if count == 0:
            print("\nInitializing database with sample products...")
            
            # SECURE: Using parameterized INSERT
            sample_products = [
                ("Laptop Pro 15", "High-performance laptop with 16GB RAM", 1299.99, "Electronics", 15),
                ("Wireless Mouse", "Ergonomic wireless mouse", 29.99, "Electronics", 50),
                ("Office Chair", "Comfortable ergonomic chair", 249.99, "Furniture", 20),
                ("Desk Lamp", "LED desk lamp with adjustable brightness", 39.99, "Furniture", 30),
                ("Python Book", "Learn Python programming", 49.99, "Books", 100),
            ]
            
            insert_query = """
                INSERT INTO products (name, description, price, category, stock_quantity)
                VALUES (%s, %s, %s, %s, %s)
            """
            
            cursor.executemany(insert_query, sample_products)
            connection.commit()
            
            print(f"✓ Created {len(sample_products)} sample products")
        
        cursor.close()
        
    except Error as e:
        print(f"✗ Error initializing database: {e}")
        connection.rollback()

def demonstrate_sql_injection():
    """
    Demonstrate SQL injection prevention.
    """
    print("\n" + "="*70)
    print("SQL INJECTION PREVENTION DEMONSTRATION")
    print("="*70)
    
    print("\n❌ INSECURE APPROACH (vulnerable):")
    print("-" * 70)
    product_id = "1"
    new_name = "Hacked'; DROP TABLE products; --"
    
    insecure_query = f"UPDATE products SET name = '{new_name}' WHERE id = {product_id}"
    print("If we built the query with string formatting:")
    print(f"  {insecure_query}")
    print("\nThis would execute:")
    print("  UPDATE products SET name = 'Hacked';")
    print("  DROP TABLE products; -- WHERE id = 1")
    print("\nThe entire products table would be DELETED!")
    
    print("\n✅ SECURE APPROACH (parameterized):")
    print("-" * 70)
    secure_query = "UPDATE products SET name = %s WHERE id = %s"
    print("Query template with placeholders:")
    print(f"  {secure_query}")
    print("Parameters passed separately:")
    print(f"  ('{new_name}', {product_id})")
    print("\nThe database treats the entire name as data, not SQL code.")
    print("The malicious SQL is stored as a literal string value.")
    print("Your database remains safe!")

def simple_update(connection):
    """
    Perform a simple product name update with full validation.
    """
    print("\n" + "="*60)
    print("UPDATE PRODUCT NAME")
    print("="*60)
    
    # Get product ID
    product_id_str = input("\nEnter Product ID: ").strip()
    
    # Validate product ID
    is_valid, product_id, error = validate_product_id(product_id_str)
    if not is_valid:
        print(f"✗ {error}")
        return
    
    # Get current product info
    print(f"\n✓ Fetching product information...")
    product = get_product_info(connection, product_id)
    
    if not product:
        print(f"✗ No product found with ID: {product_id}")
        return
    
    # Display current product info
    print("\nCurrent product information:")
    display_product(product)
    
    # Get new product name
    new_name = input("\nEnter new product name: ").strip()
    
    # Validate new name
    is_valid, cleaned_name, error = validate_product_name(new_name)
    if not is_valid:
        print(f"✗ {error}")
        return
    
    # Confirm update
    print(f"\nYou are about to change:")
    print(f"  Old name: {product['name']}")
    print(f"  New name: {cleaned_name}")
    
    confirm = input("\nConfirm update? (yes/no): ").strip().lower()
    
    if confirm not in ['yes', 'y']:
        print("✗ Update cancelled")
        return
    
    # Perform the update
    print("\n✓ Updating product...")
    rows_updated = update_product_name(connection, product_id, cleaned_name)
    
    # Report results securely
    print("\n" + "="*60)
    if rows_updated == 1:
        print("✓ UPDATE SUCCESSFUL")
        print(f"✓ {rows_updated} row was modified")
    elif rows_updated == 0:
        print("⚠ UPDATE COMPLETED")
        print("⚠ 0 rows were modified (product may not exist)")
    elif rows_updated > 1:
        print("⚠ UNEXPECTED RESULT")
        print(f"⚠ {rows_updated} rows were modified (expected 1)")
    else:
        print("✗ UPDATE FAILED")
    print("="*60)
    
    # Display updated product
    if rows_updated > 0:
        print("\nUpdated product information:")
        updated_product = get_product_info(connection, product_id)
        if updated_product:
            display_product(updated_product)

def advanced_update(connection):
    """
    Update multiple product fields securely.
    """
    print("\n" + "="*60)
    print("ADVANCED PRODUCT UPDATE")
    print("="*60)
    
    # Get product ID
    product_id_str = input("\nEnter Product ID: ").strip()
    
    is_valid, product_id, error = validate_product_id(product_id_str)
    if not is_valid:
        print(f"✗ {error}")
        return
    
    # Get current product
    product = get_product_info(connection, product_id)
    if not product:
        print(f"✗ No product found with ID: {product_id}")
        return
    
    print("\nCurrent product information:")
    display_product(product)
    
    # Collect updates
    updates = {}
    
    new_name = input("\nNew name (press Enter to skip): ").strip()
    if new_name:
        is_valid, cleaned_name, error = validate_product_name(new_name)
        if is_valid:
            updates['name'] = cleaned_name
        else:
            print(f"✗ Invalid name: {error}")
            return
    
    new_price = input("New price (press Enter to skip): ").strip()
    if new_price:
        try:
            price = float(new_price)
            if price >= 0:
                updates['price'] = price
            else:
                print("✗ Price must be non-negative")
                return
        except ValueError:
            print("✗ Invalid price format")
            return
    
    new_stock = input("New stock quantity (press Enter to skip): ").strip()
    if new_stock:
        try:
            stock = int(new_stock)
            if stock >= 0:
                updates['stock_quantity'] = stock
            else:
                print("✗ Stock must be non-negative")
                return
        except ValueError:
            print("✗ Invalid stock format")
            return
    
    if not updates:
        print("✗ No updates provided")
        return
    
    # Confirm
    print("\nFields to update:")
    for field, value in updates.items():
        print(f"  {field}: {value}")
    
    confirm = input("\nConfirm update? (yes/no): ").strip().lower()
    if confirm not in ['yes', 'y']:
        print("✗ Update cancelled")
        return
    
    # Perform update
    rows_updated = update_multiple_fields(connection, product_id, updates)
    
    print("\n" + "="*60)
    if rows_updated == 1:
        print(f"✓ Successfully updated {rows_updated} row")
        print(f"✓ Modified fields: {', '.join(updates.keys())}")
    else:
        print(f"⚠ Updated {rows_updated} rows")
    print("="*60)

def main():
    """Main function for secure product update system"""
    print("="*70)
    print("        SECURE POSTGRESQL PRODUCT UPDATE")
    print("="*70)
    print("\nThis system demonstrates secure database updates using")
    print("parameterized queries to prevent SQL injection attacks.")
    
    # Connect to database
    connection = connect_to_database()
    
    if not connection:
        print("\n✗ Cannot proceed without database connection")
        print("Please check your database configuration in DB_CONFIG")
        return
    
    try:
        # Initialize database
        init_sample_database(connection)
        
        while True:
            print("\n" + "="*60)
            print("OPTIONS:")
            print("  1. Update product name")
            print("  2. Advanced update (multiple fields)")
            print("  3. View product by ID")
            print("  4. Show SQL injection demonstration")
            print("  5. Quit")
            print("="*60)
            
            choice = input("\nSelect option: ").strip()
            
            if choice == '1':
                simple_update(connection)
            
            elif choice == '2':
                advanced_update(connection)
            
            elif choice == '3':
                product_id_str = input("\nEnter Product ID: ").strip()
                is_valid, product_id, error = validate_product_id(product_id_str)
                if is_valid:
                    product = get_product_info(connection, product_id)
                    if product:
                        display_product(product)
                    else:
                        print(f"✗ No product found with ID: {product_id}")
                else:
                    print(f"✗ {error}")
            
            elif choice == '4':
                demonstrate_sql_injection()
            
            elif choice == '5':
                print("\nGoodbye!")
                break
            
            else:
                print("✗ Invalid option")
    
    finally:
        if connection:
            connection.close()
            print("\n✓ Database connection closed")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nInterrupted by user. Goodbye!")
    except Exception as e:
        print(f"\n✗ Unexpected error: {e}")
        sys.exit(1)

File: code samples/Java/Security-Aware/Claude/prompt5.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Timestamp;
import java.util.Scanner;

public class prompt5 {
    
    // Database connection parameters - modify for your setup
    private static final String DB_URL = "jdbc:postgresql://localhost:5432/shop_db";
    private static final String DB_USER = "your_username";
    private static final String DB_PASSWORD = "your_password";
    
    // Validation constraints
    private static final int MIN_PRODUCT_NAME_LENGTH = 1;
    private static final int MAX_PRODUCT_NAME_LENGTH = 255;
    private static final int MIN_PRODUCT_ID = 1;
    private static final int MAX_PRODUCT_ID = 1000000;
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("=== Secure Product Update Tool ===");
        System.out.println("Update product information safely\n");
        
        // Get product ID
        System.out.print("Enter Product ID: ");
        String productIdInput = scanner.nextLine().trim();
        
        // Validate product ID
        int productId;
        if (!validateProductId(productIdInput)) {
            scanner.close();
            return;
        }
        productId = Integer.parseInt(productIdInput);
        
        // Get new product name
        System.out.print("Enter new product name: ");
        String newName = scanner.nextLine().trim();
        
        // Validate product name
        if (!validateProductName(newName)) {
            scanner.close();
            return;
        }
        
        scanner.close();
        
        // Show current product info before updating
        System.out.println("\n--- Current Product Information ---");
        ProductInfo currentInfo = getProductInfo(productId);
        
        if (currentInfo == null) {
            System.out.println("✗ Product ID " + productId + " not found.");
            System.out.println("Update cancelled.");
            return;
        }
        
        displayProductInfo(currentInfo);
        
        // Perform the secure update
        int rowsUpdated = updateProductName(productId, newName);
        
        // Show updated product info
        if (rowsUpdated > 0) {
            System.out.println("\n--- Updated Product Information ---");
            ProductInfo updatedInfo = getProductInfo(productId);
            if (updatedInfo != null) {
                displayProductInfo(updatedInfo);
            }
        }
    }
    
    /**
     * Validates product ID input.
     * 
     * @param productIdInput The user-provided product ID string
     * @return true if valid, false otherwise
     */
    public static boolean validateProductId(String productIdInput) {
        System.out.println("\n--- Validating Product ID ---");
        
        // Check if empty
        if (productIdInput == null || productIdInput.isEmpty()) {
            System.out.println("✗ Error: Product ID cannot be empty.");
            return false;
        }
        
        // Check if numeric
        try {
            int productId = Integer.parseInt(productIdInput);
            
            // Check range
            if (productId < MIN_PRODUCT_ID || productId > MAX_PRODUCT_ID) {
                System.out.println("✗ Error: Product ID must be between " + 
                    MIN_PRODUCT_ID + " and " + MAX_PRODUCT_ID + ".");
                return false;
            }
            
            System.out.println("✓ Product ID validation passed: " + productId);
            return true;
            
        } catch (NumberFormatException e) {
            System.out.println("✗ Error: Product ID must be a valid integer.");
            return false;
        }
    }
    
    /**
     * Validates product name input.
     * 
     * @param productName The user-provided product name
     * @return true if valid, false otherwise
     */
    public static boolean validateProductName(String productName) {
        System.out.println("\n--- Validating Product Name ---");
        
        // Check if empty
        if (productName == null || productName.isEmpty()) {
            System.out.println("✗ Error: Product name cannot be empty.");
            return false;
        }
        
        // Check length
        if (productName.length() < MIN_PRODUCT_NAME_LENGTH) {
            System.out.println("✗ Error: Product name too short (minimum " + 
                MIN_PRODUCT_NAME_LENGTH + " character).");
            return false;
        }
        
        if (productName.length() > MAX_PRODUCT_NAME_LENGTH) {
            System.out.println("✗ Error: Product name too long (maximum " + 
                MAX_PRODUCT_NAME_LENGTH + " characters).");
            return false;
        }
        
        // Check for only whitespace
        if (productName.trim().isEmpty()) {
            System.out.println("✗ Error: Product name cannot be only whitespace.");
            return false;
        }
        
        // Optional: Check for suspicious SQL keywords (for logging/monitoring)
        // Note: PreparedStatement makes this safe anyway
        String[] suspiciousPatterns = {"--", "/*", "*/", ";", "drop", "delete", "truncate"};
        String lowerName = productName.toLowerCase();
        
        for (String pattern : suspiciousPatterns) {
            if (lowerName.contains(pattern)) {
                System.out.println("⚠ Warning: Product name contains suspicious pattern: " + pattern);
                System.out.println("  (Note: This is safe with PreparedStatement)");
                break;
            }
        }
        
        System.out.println("✓ Product name validation passed: \"" + productName + "\"");
        return true;
    }
    
    /**
     * SECURE PRODUCT UPDATE
     * Updates product name using PreparedStatement with parameterized query.
     * Prevents SQL injection attacks.
     * 
     * @param productId The ID of the product to update
     * @param newName The new name for the product
     * @return Number of rows updated (should be 0 or 1)
     */
    public static int updateProductName(int productId, String newName) {
        System.out.println("\n--- Executing Secure Update ---");
        
        // SECURE: Using PreparedStatement with ? placeholders
        // This prevents SQL injection by treating input as data, not SQL code
        String sql = "UPDATE products SET name = ?, updated_at = CURRENT_TIMESTAMP WHERE product_id = ?";
        
        System.out.println("Query: UPDATE products SET name = ?, updated_at = CURRENT_TIMESTAMP WHERE product_id = ?");
        System.out.println("Parameters: [" + newName + "], [" + productId + "]");
        System.out.println("Method: PreparedStatement (SQL injection safe)");
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {
            
            // CRITICAL: Set parameters safely using setString() and setInt()
            // This prevents SQL injection
            pstmt.setString(1, newName);     // New product name
            pstmt.setInt(2, productId);       // Product ID for WHERE clause
            
            // Execute the update and get the number of affected rows
            int rowsUpdated = pstmt.executeUpdate();
            
            System.out.println("\n--- Update Result ---");
            System.out.println("Rows updated: " + rowsUpdated);
            
            if (rowsUpdated == 0) {
                System.out.println("⚠ Warning: No product found with ID " + productId);
                System.out.println("  The product may have been deleted or the ID is incorrect.");
            } else if (rowsUpdated == 1) {
                System.out.println("✓ Product updated successfully!");
            } else {
                // This should never happen with proper primary key constraint
                System.out.println("⚠ Warning: Multiple rows were updated (" + rowsUpdated + ")");
                System.out.println("  This indicates a database integrity issue.");
            }
            
            return rowsUpdated;
            
        } catch (SQLException e) {
            System.err.println("\n✗ Database Error: " + e.getMessage());
            System.err.println("\nTroubleshooting:");
            System.err.println("1. Verify PostgreSQL server is running");
            System.err.println("2. Check database connection parameters");
            System.err.println("3. Ensure database and table exist");
            System.err.println("4. Verify PostgreSQL JDBC driver is in classpath");
            System.err.println("5. Check user permissions for UPDATE operation");
            return -1;
        }
    }
    
    /**
     * Retrieves current product information.
     * Uses PreparedStatement for security.
     * 
     * @param productId The product ID to retrieve
     * @return ProductInfo object or null if not found
     */
    public static ProductInfo getProductInfo(int productId) {
        // SECURE: Using PreparedStatement
        String sql = "SELECT product_id, name, description, price, stock_quantity, category, " +
                    "created_at, updated_at FROM products WHERE product_id = ?";
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {
            
            pstmt.setInt(1, productId);
            ResultSet rs = pstmt.executeQuery();
            
            if (rs.next()) {
                ProductInfo info = new ProductInfo();
                info.productId = rs.getInt("product_id");
                info.name = rs.getString("name");
                info.description = rs.getString("description");
                info.price = rs.getDouble("price");
                info.stockQuantity = rs.getInt("stock_quantity");
                info.category = rs.getString("category");
                info.createdAt = rs.getTimestamp("created_at");
                info.updatedAt = rs.getTimestamp("updated_at");
                return info;
            }
            
            return null;
            
        } catch (SQLException e) {
            System.err.println("Error retrieving product info: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Displays product information in a formatted way.
     */
    private static void displayProductInfo(ProductInfo info) {
        System.out.println("Product ID:    " + info.productId);
        System.out.println("Name:          " + info.name);
        System.out.println("Description:   " + (info.description != null ? info.description : "N/A"));
        System.out.println("Price:         $" + String.format("%.2f", info.price));
        System.out.println("Stock:         " + info.stockQuantity);
        System.out.println("Category:      " + (info.category != null ? info.category : "N/A"));
        System.out.println("Created:       " + info.createdAt);
        System.out.println("Last Updated:  " + info.updatedAt);
    }
    
    /**
     * Class to hold product information.
     */
    private static class ProductInfo {
        int productId;
        String name;
        String description;
        double price;
        int stockQuantity;
        String category;
        Timestamp createdAt;
        Timestamp updatedAt;
    }
    
    /**
     * Sets up the database with table and sample data (for testing).
     */
    public static void setupDatabase() {
        System.out.println("=== Database Setup ===\n");
        
        // Create products table
        String createTableSQL = "CREATE TABLE IF NOT EXISTS products (" +
            "product_id SERIAL PRIMARY KEY, " +
            "name VARCHAR(255) NOT NULL, " +
            "description TEXT, " +
            "price NUMERIC(10, 2) NOT NULL, " +
            "stock_quantity INTEGER NOT NULL DEFAULT 0, " +
            "category VARCHAR(100), " +
            "created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP, " +
            "updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP)";
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             Statement stmt = conn.createStatement()) {
            
            stmt.execute(createTableSQL);
            System.out.println("✓ Products table created");
            
        } catch (SQLException e) {
            System.err.println("✗ Error creating table: " + e.getMessage());
            return;
        }
        
        // Insert sample products using PreparedStatement
        String insertSQL = "INSERT INTO products (name, description, price, stock_quantity, category) " +
                          "VALUES (?, ?, ?, ?, ?)";
        
        String[][] sampleProducts = {
            {"Laptop", "High-performance laptop", "999.99", "25", "Electronics"},
            {"Mouse", "Wireless ergonomic mouse", "29.99", "150", "Accessories"},
            {"Keyboard", "Mechanical keyboard", "89.99", "75", "Accessories"},
            {"Monitor", "27-inch 4K display", "399.99", "30", "Electronics"},
            {"Headset", "Gaming headset", "79.99", "120", "Audio"}
        };
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement pstmt = conn.prepareStatement(insertSQL)) {
            
            for (String[] product : sampleProducts) {
                pstmt.setString(1, product[0]);
                pstmt.setString(2, product[1]);
                pstmt.setDouble(3, Double.parseDouble(product[2]));
                pstmt.setInt(4, Integer.parseInt(product[3]));
                pstmt.setString(5, product[4]);
                pstmt.executeUpdate();
            }
            
            System.out.println("✓ Sample products inserted");
            System.out.println("\n✓ Setup complete!");
            System.out.println("\nYou can now update products by ID (1-5)");
            
        } catch (SQLException e) {
            System.err.println("✗ Error inserting products: " + e.getMessage());
        }
    }
    
    /**
     * Demonstrates SQL injection prevention.
     */
    public static void demonstrateSQLInjection() {
        System.out.println("\n=== SQL Injection Prevention Demo ===\n");
        
        String maliciousName = "Laptop', price = 0.01 WHERE product_id > 0; --";
        int productId = 1;
        
        System.out.println("Malicious product name: " + maliciousName);
        System.out.println("Product ID: " + productId);
        
        System.out.println("\nVULNERABLE approach (DON'T DO THIS):");
        System.out.println("String sql = \"UPDATE products SET name = '\" + name + \"' WHERE product_id = \" + id;");
        
        System.out.println("\nResulting query:");
        System.out.println("UPDATE products SET name = 'Laptop', price = 0.01 WHERE product_id > 0; --' WHERE product_id = 1");
        
        System.out.println("\nWhat happens:");
        System.out.println("1. Sets name to 'Laptop'");
        System.out.println("2. Sets price to $0.01 for ALL products (product_id > 0)");
        System.out.println("3. Comments out the original WHERE clause");
        System.out.println("Result: ALL products now cost $0.01! Massive financial loss!");
        
        System.out.println("\n" + "=".repeat(70));
        System.out.println("\nSECURE approach (OUR CODE):");
        System.out.println("String sql = \"UPDATE products SET name = ? WHERE product_id = ?\";");
        System.out.println("pstmt.setString(1, name);");
        System.out.println("pstmt.setInt(2, id);");
        
        System.out.println("\nHow PreparedStatement handles it:");
        System.out.println("Parameter 1: Laptop', price = 0.01 WHERE product_id > 0; --");
        System.out.println("Parameter 2: 1");
        
        System.out.println("\nWhat happens:");
        System.out.println("The entire malicious string becomes the product name!");
        System.out.println("Only product ID 1 is updated with that weird name.");
        System.out.println("No price changes, no other products affected.");
        System.out.println("Result: SQL injection prevented! ✓");
    }
    
    /**
     * Batch update example (updates multiple products safely).
     */
    public static void batchUpdateExample() {
        System.out.println("\n=== Batch Update Example ===\n");
        
        String sql = "UPDATE products SET price = price * ? WHERE category = ?";
        
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {
            
            // 10% discount on Electronics
            pstmt.setDouble(1, 0.9);
            pstmt.setString(2, "Electronics");
            int updated = pstmt.executeUpdate();
            
            System.out.println("✓ Applied 10% discount to " + updated + " Electronics products");
            
        } catch (SQLException e) {
            System.err.println("Error in batch update: " + e.getMessage());
        }
    }
}

/*
SECURITY FEATURES:

✓ PreparedStatement with parameterized queries
✓ Comprehensive input validation
✓ Type-safe parameter binding
✓ No string concatenation in SQL
✓ Before/after comparison
✓ Proper error handling
✓ Row count verification
✓ SQL injection prevention

HOW PARAMETERIZED UPDATES PREVENT SQL INJECTION:

VULNERABLE CODE (DON'T DO THIS):
```java
String sql = "UPDATE products SET name = '" + newName + 
             "' WHERE product_id = " + productId;
Statement stmt = conn.createStatement();
int rows = stmt.executeUpdate(sql);  // DANGEROUS!
```

Attack input:
Product ID: 1
New name: Laptop', price = 0.01 WHERE product_id > 0; --

Resulting query:
UPDATE products SET name = 'Laptop', price = 0.01 WHERE product_id > 0; --' WHERE product_id = 1

What happens:
1. Sets name to 'Laptop'
2. Sets price to $0.01 for ALL products
3. Comments out the intended WHERE clause
Result: ALL products now cost $0.01! Financial catastrophe!

SECURE CODE (OUR APPROACH):
```java
String sql = "UPDATE products SET name = ? WHERE product_id = ?";
PreparedStatement pstmt = conn.prepareStatement(sql);
pstmt.setString(1, newName);
pstmt.setInt(2, productId);
int rows = pstmt.executeUpdate();
```

Same attack input:
PreparedStatement treats the entire string as the name value:
"Laptop', price = 0.01 WHERE product_id > 0; --"

Result:
Only product ID 1 gets updated with that literal string as its name.
No SQL injection, no price changes to other products!

SQL INJECTION ATTACKS PREVENTED:

1. Price Manipulation:
   Input: Laptop', price = 0 --
   Result: Treated as literal product name

2. Mass Updates:
   Input: X' WHERE 1=1; --
   Result: Treated as literal product name

3. Data Deletion:
   Input: X'; DELETE FROM products; --
   Result: Treated as literal product name, no deletion

4. Privilege Escalation:
   Input: X'; UPDATE users SET role='admin'; --
   Result: Treated as literal product name

VALIDATION FEATURES:

1. Product ID Validation:
   ✓ Must be numeric
   ✓ Range check (1 to 1,000,000)
   ✓ Type-safe integer parsing

2. Product Name Validation:
   ✓ Non-empty check
   ✓ Length constraints (1-255 chars)
   ✓ Whitespace-only rejection
   ✓ Suspicious pattern detection

3. Update Verification:
   ✓ Shows current state before update
   ✓ Reports rows affected
   ✓ Shows updated state after update
   ✓ Warns if no rows updated

SETUP INSTRUCTIONS:

1. Install PostgreSQL JDBC Driver:
   Maven:
   <dependency>
       <groupId>org.postgresql</groupId>
       <artifactId>postgresql</artifactId>
       <version>42.6.0</version>
   </dependency>

2. Create PostgreSQL database:
   CREATE DATABASE shop_db;

3. Update connection parameters:
   - DB_URL: your database URL
   - DB_USER: your PostgreSQL username
   - DB_PASSWORD: your PostgreSQL password

4. Setup database (run once):
   Uncomment: setupDatabase();

5. Compile and run:
   javac SecureProductUpdater.java
   java -cp .:postgresql.jar SecureProductUpdater

USAGE EXAMPLE:

=== Secure Product Update Tool ===

Enter Product ID: 1

--- Validating Product ID ---
✓ Product ID validation passed: 1

Enter new product name: Gaming Laptop

--- Validating Product Name ---
✓ Product name validation passed: "Gaming Laptop"

--- Current Product Information ---
Product ID:    1
Name:          Laptop
Price:         $999.99
Stock:         25
Created:       2024-01-15 10:30:00
Last Updated:  2024-01-15 10:30:00

--- Executing Secure Update ---
Query: UPDATE products SET name = ?, updated_at = CURRENT_TIMESTAMP WHERE product_id = ?
Parameters: [Gaming Laptop], [1]
Method: PreparedStatement (SQL injection safe)

--- Update Result ---
Rows updated: 1
✓ Product updated successfully!

--- Updated Product Information ---
Product ID:    1
Name:          Gaming Laptop
Price:         $999.99
Stock:         25
Created:       2024-01-15 10:30:00
Last Updated:  2024-11-29 14:45:23

PRODUCTION ENHANCEMENTS:

1. Transaction support for complex updates
2. Optimistic locking (version numbers)
3. Audit trail logging
4. Change history tracking
5. Role-based access control
6. Concurrent update detection
7. Rollback capabilities
8. Notification system for changes

The fundamental principle: Always use PreparedStatement with
parameterized queries for ANY database modification!
*/